apps/system/update/binary_update.c

```diff
@@ -0,0 +1,237 @@
+		return ERROR;
+	}
+
+	memset((void *)&request_msg, 0, sizeof(binmgr_request_t));
+
+	request_msg.cmd = BINMGR_GET_INFO;
+	request_msg.pid = getpid();
+	snprintf(request_msg.q_name, BIN_PRIVMQ_LEN, "%s%d", BINMGR_RESPONSE_MQ, request_msg.pid);
+	snprintf(request_msg.b_name, BIN_NAME_MAX, "%s", binary_name);
+
+	ret = send_requset(&request_msg);
+	if (ret == ERROR) {
+		dbg("mq_send failed!\n");
+		return ERROR;
+	}
+
+	attr.mq_maxmsg = BINMGR_MAX_MSG;
+	attr.mq_msgsize = sizeof(binmgr_response_t);
+	attr.mq_flags = 0;
+
+	private_mq = mq_open(request_msg.q_name, O_RDONLY | O_CREAT, 0666, &attr);
+
+	if (private_mq == (mqd_t)ERROR) {
+		dbg("mq_open failed!\n");
+		return ERROR;
+	}
+
+	ret = receive_response(private_mq, &response_msg);
+	if (ret == ERROR) {
+		mq_close(private_mq);
+		mq_unlink(request_msg.q_name);
+		return ERROR;
+	}
+
+	mq_close(private_mq);
+	mq_unlink(request_msg.q_name);
+
+	if (response_msg.cmd == REQUESET_NOT_INFO) {
+		dbg("There is no binary information.!\n");
+		return ERROR;
+	}
+
+	binary_info_copy(binary_info, &response_msg);
+
+	return OK;
+}
+
+int get_binary_info_all(binary_info_list_t *binary_info_list)
+{
+	int ret;
+	mqd_t private_mq;
+
+	binmgr_request_t request_msg;
+	binmgr_response_t response_msg;
+	binary_info_list_t *next_list;
+
+	struct mq_attr attr;
+
+	memset((void *)&request_msg, 0, sizeof(binmgr_request_t));
+
+	request_msg.cmd = BINMGR_GET_INFO_ALL;
+	request_msg.pid = getpid();
+
+	snprintf(request_msg.q_name, BIN_NAME_MAX, "%s%d", BINMGR_RESPONSE_MQ, request_msg.pid);
+
+	send_requset(&request_msg);
+
+	attr.mq_maxmsg = BINMGR_MAX_MSG;
+	attr.mq_msgsize = sizeof(binmgr_response_t);
+	attr.mq_flags = 0;
+
+	private_mq = mq_open(request_msg.q_name, O_RDONLY | O_CREAT, 0666, &attr);
+	if (private_mq == (mqd_t)ERROR) {
+		dbg("mq_open failed!\n");
+		return ERROR;
+	}
+
+	ret = receive_response(private_mq, &response_msg);
+	if (ret == ERROR) {
+		mq_close(private_mq);
+		mq_unlink(request_msg.q_name);
+		return ERROR;
+	}
+
+	binary_info_copy(&(binary_info_list->binary_info), &response_msg);
+
+	next_list = binary_info_list;
+	if (response_msg.cmd == REQUESET_NOT_INFO) {
+		dbg("There is no binary information.!\n");
+		mq_close(private_mq);
+		mq_unlink(request_msg.q_name);
+		return ERROR;
+	}
+
+	while (response_msg.cmd == REQUESET_CONTINUE) {
+		ret = receive_response(private_mq, &response_msg);
+		if (ret == ERROR) {
+			binary_list_clean(binary_info_list);
+			mq_close(private_mq);
+			mq_unlink(request_msg.q_name);
+			return ERROR;
+		}
+		next_list->next = (binary_info_list_t *)malloc(sizeof(binary_info_list_t));
+		next_list = next_list->next;
+		next_list->next = NULL;
+		binary_info_copy(&(next_list->binary_info), &response_msg);
+	}
+
+	mq_close(private_mq);
+	mq_unlink(request_msg.q_name);
+	return OK;
+}
+
+void binary_list_clean(binary_info_list_t *binary_info_list)
+{
+	if (binary_info_list->next != NULL) {
+		binary_list_clean(binary_info_list->next);
```

Comment:
Instead of calling recursively, the list can be traversed linearly and cleaned up.